Experimental AVR target improvements

gergoerdi opened the pull request:
Noob alert: **Please review very carefully** -- I haven't even looked at LLVM's internals until a week ago.

The commits should be fairly self-explanatory; the fixes are for the following issues:

* https://github.com/avr-rust/rust/issues/44
* https://github.com/avr-rust/rust/issues/46
* https://github.com/avr-rust/rust/issues/47

gergoerdi committed:
[AVR] Add pseudo-instructions for relaxed branches.
For now, these relaxed branches are used unconditionally, even when
the branch target is nearby.
https://github.com/avr-rust/rust/issues/44

gergoerdi committed:
[AVR] Implement LPMWRdZ pseudo-instruction's expansion.
FIXME: implementation is mostly copy-pasted from LDWRdPtr, so we should
refactor a bit and unify the two

shepmaster commented:
Thanks for the patches! I think our preference is to land everything in upstream LLVM before it finds its way into this repository, but this seems like a good place to track those patches *before* then (and also to allow @dylanmckay to grab them and review them)

Said another way: I think this PR can stay open, but not be merged until each of the commits has been merged into LLVM.

gergoerdi commented:
Oh, I thought it was exactly the other way round -- that this repo is to stage dubious changes until they can be upstreamed to LLVM.

BTW, most definitely don't merge this PR, since the last three commits break structure unpacking: https://github.com/avr-rust/rust/issues/48

gergoerdi committed:
[AVR] Add ISel pattern which maches load-from-global, compile it to
LPMRdZ (https://github.com/avr-rust/rust/issues/47,
https://github.com/avr-rust/rust/issues/48)

gergoerdi committed:
[AVR] Add 16-bit version of load-from-global

gergoerdi committed:
[AVR] Use separate ProgMemWrapper for GlobalAddresses of constants

gergoerdi committed:
#TEMP add temporary instrument selector pattern for creating pointers
to ProgMem

gergoerdi commented:
This branch now has a ton of useful fixes with which I was able to compile a non-trivial Rust app to AVR (I hope to write a blog post about it shortly). Can I leave it in your hands to upstream it to LLVM proper?

shepmaster commented:
@gergoerdi each separate fix to LLVM should have one or more tests associated with it; otherwise it's highly likely that any fixes will regress with time. 

@dylanmckay any other things you'd like to see?

dylanmckay commented on lib/Target/AVR/AVRInstrInfo.td:
```diff
@@ -2057,6 +2109,10 @@ def : Pat<(store i8:$src, (i16 (AVRWrapper tglobaladdr:$dst))),
 def : Pat<(store i16:$src, (i16 (AVRWrapper tglobaladdr:$dst))),
           (STSWKRr tglobaladdr:$dst, i16:$src)>;
 
+// Temporary workaround -- this one is BOGUS!
```
Why is this temporary? What is it working around? Can you pop a comment somewhere with an explanation?

gergoerdi commented:
@shepmaster OK I'll try to find the time to add tests for the new fixes. My question still stands though -- can you or @dylanmckay take care of the upstreaming? I'm not yet invested enough in this to want to take on that hassle.

gergoerdi commented:
Also, I think there's three categories of my commits here:

1. Definite bugfixes, no reason not to merge them:

- [x] https://github.com/avr-rust/llvm/pull/1/commits/c7d5217f866a3e07bb9dd2b337b61463356dea77 (avr-rust/rust#49)
- [x] https://github.com/avr-rust/llvm/pull/1/commits/9fe62d7238036defc258118a2f2b74127fb771f1 (avr-rust/rust#50)
- [x] https://github.com/avr-rust/llvm/pull/1/commits/910429cca6d7b1bc9cd0d2f22f1bd1d011c54fee (avr-rust/rust#52)

2. Definite improvement, but implementation could be better

- [ ] https://github.com/avr-rust/llvm/pull/1/commits/383e429f720ab50122bb7adcd22aabcc1e361671 (avr-rust/rust#47)

3. Desperate struggling piling hacks upon hacks fighting the Harvard architecture

- [ ] https://github.com/avr-rust/llvm/pull/1/commits/de2539285a6ab0287466cbe9a79a8757dba964f3 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/07a96e39d25cd2db8b710a826e941bb5aa7b0a16 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/bff3af337679c389c755b53ab826f2816f18dae4 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/65f3f2abb64c23fe29aacdd602d5346d388945b1  (avr-rust/rust#52)

I don't think commits from the third category should be merged; I'll open a separate ticket where we can hopefully come up with a proper solution to `PROGMEM`-related issues.

dylanmckay commented:
I'm happy to take on upstreaming @gergoerdi, will keep you updated.

Commit:
[AVR] When lowering Select8/Select16, put newly generated MBBs in the same spot

Contributed by Dr. Gergő Érdi.

Fixes a bug.

Raised from (https://github.com/avr-rust/rust/issues/49).

git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@302973 91177308-0d34-0410-b5e6-96231b3b80d8

gergoerdi committed:
[AVR] When lowering shifts into loops, put newly generated MBBs in the same
spot as the original MBB (https://github.com/avr-rust/rust/issues/62)

gergoerdi committed:
[AVR] Elaborate LDWRdPtr into `ld r, X++; ld r+1, X`
https://github.com/avr-rust/rust/issues/58

gergoerdi committed:
[AVR] Use 1-byte alignment for all types
(https://github.com/avr-rust/rust/issues/63 https://github.com/avr-rust/rust/issues/64)

[AVR] Use 1-byte alignment for all data types
(https://github.com/avr-rust/rust/issues/64)

Commit:
[AVR] Remove SREG from CPI's Uses; authored by Florian Zeitz

Summary: CPI does not read the status register, but only writes it.

Reviewers: dylanmckay

Reviewed By: dylanmckay

Subscribers: llvm-commits

Differential Revision: https://reviews.llvm.org/D33223

git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@304116 91177308-0d34-0410-b5e6-96231b3b80d8

dylanmckay commented:
I'm going to put this comment here to track which commits I have upstreamed

Commits upstreamed

* [AVR] Add pseudo-instructions for relaxed branches.

Commit:
[AVR] CPIRdK can only work with r16..r31; Authored by Dr. Gergo Erdi
(https://github.com/avr-rust/rust/issues/50)

git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@304283 91177308-0d34-0410-b5e6-96231b3b80d8

Commit:
[AVR] Fix a big in shift operator lowering; Authored by Dr. Gergo Erdi

When generating code for a shift loop, check the shift
 amount against the literal value 0, not R0

git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@304284 91177308-0d34-0410-b5e6-96231b3b80d8

dylanmckay commented:
It looks like there's some stuff here that still needs to be merged

I will change the base branch to see more clearly what is missing that should go into master

dylanmckay commented:
That didn't help, the histories are different!

dylanmckay commented:
Stuff to merge

* "[AVR] Elaborate LDWRdPtr into `ld r, X++; ld r+1, X`"
* Grab `mayLoad` change from the diff out of "[AVR] Add ISel pattern which maches load-from-global, compile it to LPMRdZ"
* "[AVR] Implement LPMWRdZ pseudo-instruction's expansion."

I believe everything else is already upstreamed, or were temporary hacks

dylanmckay commented:
Ran tests locally, all pass

dylanmckay commented:
Wrong merge request! did not mean to merge, nor did I run tests for this PR.

Reopening now, have reverted

dylanmckay commented:
Looks like there's no reopen button, will just cherry-pick stuff now

dylanmckay commented:
Upstreamed ""[AVR] Elaborate LDWRdPtr into ld r, X++; ld r+1, X"" in [r314896](http://llvm.org/viewvc/llvm-project?view=revision&revision=314896).

I also fixed a bunch of tests before committing

dylanmckay commented:
Upsteamed `mayLoad` factoring out in [r314897](http://llvm.org/viewvc/llvm-project?view=revision&revision=314897)

dylanmckay commented:
Upsteamed "[AVR] Implement LPMWRdZ pseudo-instruction's expansion." in [r314898](http://llvm.org/viewvc/llvm-project?view=revision&reision=314898)

dylanmckay commented:
Cherry-picked the upstreamed commits to avr-rust/llvm in

06d0f6b412b [AVR] Implement LPMWRdZ pseudo-instruction's expansion.
f6d00a7140b [AVR] Factor out mayLoad in tablegen patterns
d4274bd15fc [AVR] Elaborate LDWRdPtr into `ld r, X++; ld r+1, X`

dylanmckay commented:
Updated version of LLVM in for Rust in https://github.com/avr-rust/rust/commit/a22dd4652b512598bb2783740abc70d98296a16f

dylanmckay commented:
Added these commits to the current AVR 5.0.1 merge request [PR34829](https://bugs.llvm.org/show_bug.cgi?id=34829)

dylanmckay commented:
Have double checked all the commits in this PR and everything looks good.

Thanks for the patches @gergoerdi, sorry for taking so long, completely forgot about this PR!
